app/commands/v2/patch_link_set.rb

```diff
@@ -18,17 +18,25 @@ def call
         lock_version.save!
 
         grouped_links.each do |group, payload_content_ids|
+          # For each set of links in a LinkSet scoped by link_type, this iterator
+          # deletes the entire existing set and then imports all the links in the
+          # payload, preserving their ordering. The activerecord-import gem is
+          # used to create a single SQL statement for importing all links.
           links = link_set.links.where(link_type: group)
           existing_content_ids = links.pluck(:target_content_id)
 
-          content_ids_to_create = payload_content_ids - existing_content_ids
-          content_ids_to_delete = existing_content_ids - payload_content_ids
+          links.where(target_content_id: existing_content_ids).delete_all
```

Comment:
I agree with this approach but just wanted to note that there is still a separate insert for each link at the moment. So I'd expect this to perform roughly the same in the case where you're inserting a set of links for the first time, but if the user makes minor changes (adding a new tag to a long list) you lose the optimisation of just inserting the one you changed. 

That could maybe be remedied by inserting all the rows in one sql statement but I'm not sure you can do it with plain activerecord.

It might be worth checking for noops before doing anything, because publishing apps tend to send all their links when publishing a draft.

Have you run any of the benchmarks yet?